### ISSUE TITLE
SHAP values not summing up to predictions when using predict(..., predcontrib = TRUE) in R package

### ISSUE THREAD

**Issue by efh0888** [dmlc/xgboost#7790]
I've built 2 different xgboost classification models on (more or less) the same population of interest using the caret package. For one, the SHAP values produced match the predictions exactly, whereas for the other (with higher AUC) they do not match.

First model where SHAP values match the predictions (once converting to log-odds scale):

```
xgb_caret <- readRDS("xgb_caret_1.rds")
prob_good <- predict(xgb_caret, type = "prob", sparse_feature_matrix)
log_odds_good <- log(prob_good / (1 - prob_good))
summary(tibble(log_odds_good = log_odds_good$GOOD))
```
![image](https://user-images.githubusercontent.com/7320182/162246348-3dfe29e9-d324-45e3-a468-2e2103d11aed.png)

```
shap_contrib <- predict(xgb_caret$finalModel, sparse_feature_matrix, predcontrib = TRUE)
summary(shap_contrib %>% 
          as_tibble() %>% 
          mutate(row_sum_log_odds = rowSums(across(where(is.numeric)))))
```
![image](https://user-images.githubusercontent.com/7320182/162246495-52682b97-51d7-4bc5-8b47-cc6222d51e60.png)

However, running the *same exact code as above* only with a different xgboost model (also built with caret) I get a huge difference:
![image](https://user-images.githubusercontent.com/7320182/162246667-78258dd8-0b49-4c69-86d0-077fe6be29a9.png)
![image](https://user-images.githubusercontent.com/7320182/162246886-52bf0015-15e4-45cd-a735-4970ff430d13.png)

Are there any valid reasons why they may not match? Or is this in all likelihood the result some problem? I unfortunately cannot provide a reprex as I'm working with proprietary company data.

My version of xgboost is a bit older: 1.4.1.1

**Comment by kransom14** [comment]
What are the differences in the training parameters between your two models? I've had problems with the shap values not adding up to the predictions when using alternative [loss functions ](https://github.com/dmlc/xgboost/issues/7448), or when including [weights](https://github.com/slundberg/shap/issues/2395) on my observations. Unfortunately, I haven't been able to resolve either of these cases and am not sure what the solution may be.

**Comment by hcho3** [comment]
Yes, currently, the SHAP values are computed in terms of marginal units. So the correct reference is `predict(..., outputmargin = TRUE)`.

**Comment by efh0888** [comment]
@kransom14 I'm using the standard loss function and no weights. Oddly enough it appears to be an issue with caret. I realized that one difference was the original model was built on a full tuning grid of parameters whereas the second model I built just using the optimal parameters from a prior run (to save on computing time while in development). Sure enough, when I re-build the second model even with just a 2-row tuning grid, I get the correct SHAP values. Not sure what's going on under the hood, but this is a strange one for sure.

TL;DR when building an xgboost model with caret, don't ever use a 1-row tuning grid...

**Comment by efh0888** [comment]
@hcho3 so that would provide the SHAP values on the probability scale rather than the logistic scale?
